File: 107-binary-tree-level-order-traversal-ii/107-binary-tree-level-order-traversal-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        
        if ( !root )
            return {};
        
        std::vector<vector<int>> result;
        std::stack<vector<int>> stackRes;
        std::queue<TreeNode *> q;
        q.push( root );
        
        while ( q.empty() == false ) {
            int lSize = q.size();
            std::vector<int> thisLevel(0, lSize);

            
            for ( int i = 0; i < lSize; i++ ) {
                TreeNode *tmp = q.front();
                q.pop();
                thisLevel.push_back( tmp->val );
                if(tmp->left)
                    q.push( tmp->left );
                if(tmp->right)
                    q.push( tmp->right );
            }
            
            stackRes.push( thisLevel );
        }
        
        while(stackRes.empty() == false) {
            result.push_back(stackRes.top());
            stackRes.pop();
        }
        
        return result;
    }
};;

File: 107-binary-tree-level-order-traversal-ii/107-binary-tree-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        
        if root is None:
            return []
        result = collections.deque()
        q = collections.deque([root])
        
        while len(q) != 0:
            numNodes = len(q)
            temp = []
            for _ in range(numNodes):
                node = q.popleft()
                temp.append(node.val)
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
                    
            result.appendleft(temp)
        
        return result
            